slatedb/src/merge_operator.rs

```diff
@@ -119,6 +166,32 @@ pub(crate) struct MergeOperatorIterator<T: KeyValueIterator> {
     now: i64,
 }
 
+#[derive(Debug, Clone)]
+struct MergeTracker {
+    max_create_ts: Option<i64>,
+    min_expire_ts: Option<i64>,
+    seq: u64,
+}
+
+impl MergeTracker {
+    fn update(&mut self, entry: &RowEntry) -> Result<(), SlateDBError> {
+        self.max_create_ts = merge_options(self.max_create_ts, entry.create_ts, i64::max);
+        self.min_expire_ts = merge_options(self.min_expire_ts, entry.expire_ts, i64::min);
+
+        // sequence numbers should be descending
+        if self.seq < entry.seq {
+            error!(
+                "Invalid sequence number ordering: {} > {}",
+                self.seq, entry.seq
+            );
+            return Err(SlateDBError::InvalidDBState);
```

Comment:
Nit: I think a new SlateDBError that contains the error! message above (and remove error!() call) is the pattern we've been following. I worry a bit that InvalidDbState is being abused, and treated like an Error rather than SlateDBError.